// @ym/bot-engine/models/sendData.js
/**
 * Created by biddwan on 25 May 2018.
 */
const config = require('../config');
const axios = require('axios');
let controllerInstance = axios.create({
    baseURL: config.urls.CONTROLLER_SERVICE,
    headers: {"x-auth-token": process.env['apiKey']}
});
const facebookSignature = config.FACEBOOK_SIGNATURE;
const sendResponse = (bot, sender, source, data, mapping, responseType) => {
    if (source === 'facebook') {
        return controllerInstance.post('/facebook/send', {
            data: data,
            signature: facebookSignature,
            sender: sender,
            bot: bot,
            messageType: 'BOT',
            mapping: mapping
        })
    } else if (source === 'botframework' || source === 'skype') {
        return controllerInstance.post(`/botframework/send`, {
            data: data,
            sender: sender,
            bot: bot,
            messageType: 'BOT',
            mapping: mapping
        })
    } else if (source === 'yellowmessenger') {
        return controllerInstance.post('/yellowmessenger/send', {
            data: data,
            sender: sender,
            bot: bot,
            messageType: 'BOT'
        })
    } else if (source === 'sms') {
        return controllerInstance.post('/sms/send', {
            data: data,
            sender: sender,
            bot: bot,
            messageType: 'BOT',
            mapping: mapping
        })
    }
};

const sendTyping = (bot, sender) => {
    return controllerInstance.post('/yellowmessenger/send-typing', {
        sender: sender,
        bot: bot
    })
};

module.exports = {
    sendResponse,
    sendTyping
};
